modules/@apostrophecms/area/ui/apos/components/AposAreaWidget.vue

```diff
@@ -440,12 +459,109 @@ export default {
       // focus.
       apos.bus.$emit('widget-focus', { _id: this.widget._id });
     }
+
+    this.$nextTick(() => {
+      this.adminBarHeight = window.apos.adminBar.height;
+      this.controlsHeight = this.$refs.modifyControls.getBoundingClientRect().height;
+      window.addEventListener('scroll', this.stickyControlsScroll);
+      window.addEventListener('resize', this.stickyControlsResize);
+    });
+
   },
   unmounted() {
     // Remove the focus parent listener when unmounted
     apos.bus.$off('widget-focus-parent', this.focusParent);
+    window.removeEventListener('scroll', this.stickyControlsScroll);
+    window.removeEventListener('resize', this.stickyControlsResize);
   },
   methods: {
+    stickyControlsResize() {
+      if (!this.resizeTicking) {
+        this.resizeTicking = true;
+        requestAnimationFrame(() => {
+          const widgetRect = this.$refs.wrapper.getBoundingClientRect();
+
+          // If widget shifts more than tolerable don't try to track it
+          if (Math.abs(this.lastResizeTop - widgetRect.top) > this.shiftTolerance) {
+            this.stickyControlsStyles = { ...this.defaultControlsStyles };
+          } else {
+            if (this.stickyControlsStyles.position === 'fixed') {
+              this.stickyControlsStyles.right =
+                `${window.innerWidth - (widgetRect.left + widgetRect.width) + this.controlsMargin}px`;
+            }
+          }
+
+          this.lastResizeTop = widgetRect.top;
+          this.resizeTicking = false;
+        });
+      }
+    },
+    stickyControlsScroll() {
+      if (!this.scrollTicking) {
+        window.requestAnimationFrame(() => {
+          const widgetRect = this.$refs.wrapper.getBoundingClientRect();
+          const windowHeight = window.innerHeight;
+          const visibleHeight =
+            Math.min(widgetRect.bottom, windowHeight) - Math.max(widgetRect.top, 0);
+          let newStyles;
+
+          // Controls height is within 15% of the height of the widget
+          const controlsTooTallRelative = this.controlsHeight / widgetRect.height > 0.85;
+
+          // Controls are taller than widget
+          const controlsExceedWidget = widgetRect.height < this.controlsHeight;
+
+          // Repositioning in these cases feels unexpected
+          if (controlsTooTallRelative || controlsExceedWidget) {
+            return;
+          }
+
+          // Widget is under admin bar
+          if (widgetRect.top <= this.adminBarHeight) {
+
+            // Widget bottom is approaching admin bar,
+            // position controls absolutely to the bottom
+            if (
+              visibleHeight <=
+                this.controlsHeight + this.adminBarHeight + (this.controlsMargin * 3)
```

Comment:
Could store the logic into variables to avoid the need for comments.
Would help follow the if else flow too